### PR TITLE
Cfae rest

### PR DESCRIPTION
**Pull request by Casperfrc**
Creating the base API.

Right now it starts up on your local url:
0.0.0.0:1337

Right now the main issue is json decoding of the request.

### CODE REVIEW

**Comment by technoten** [comment]
Docker has been integrated nicely now, starting rest api upon container startup. Some code cleanup was done and the problem with relative paths in package imports has been solved by requiring the project to be started from the __main__.py in the root directory, see docker-compose.yml for an example of starting the project. Pytest should be run in the project root as well to ensure correct paths.

File: src/bpm/bpm_extractor.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import sys
 import os
 from multiprocessing import Pool
-from tabulate import tabulate
 
+from tabulate import tabulate
```
Comment (EmilHenriksen) [inline src/bpm/bpm_extractor.py]: move up? pep8

File: src/classification/__main__.py [hunk 2]
```diff
@@ -2,24 +2,25 @@
...
-import get_song_id as s_id
-# End of importing the utilities module
-
-from extractor.low_level_data_extractor import make_low_level_data_file
-from classifier.profile_data_extractor import get_classifier_data
-
 from pprint import pprint
 from tabulate import tabulate
-
 from multiprocessing import Pool
```
Comment (EmilHenriksen) [inline src/classification/__main__.py]: pep8 imports

File: src/classification/__main__.py [hunk 3]
```diff
@@ -2,24 +2,25 @@
...
 
+import utilities.get_song_id as s_id
+from extractor.low_level_data_extractor import make_low_level_data_file
+from classifier.profile_data_extractor import get_classifier_data
+
 
 def process_data_and_extract_profiles(song_id, song_file, song_output_file):
+    # Create temp
```
Comment (EmilHenriksen) [inline src/classification/__main__.py]: useless comment

File: src/classification/__main__.py [hunk 4]
```diff
@@ -2,24 +2,25 @@
...
+
+    #use temp
     make_low_level_data_file(song_file, song_output_file)
 
+    #use temp
```
Comment (EmilHenriksen) [inline src/classification/__main__.py]: same comment as the other use temp, what does that mean, either remove it or make it more explanitory

File: docker-compose.yml [hunk 5]
```diff
@@ -6,5 +6,7 @@ services:
     build: .
     volumes:
     - ./src:/code
```
Comment (QuincySharp) [inline docker-compose.yml]: Is this under volumes or codebase?

File: src/classification/extractor/low_level_data_extractor.py [hunk 6]
```diff
@@ -1,17 +1,21 @@
 import sys
 import subprocess
+import os
 
-from utilities.filehandler.handle_path import get_absolute_path
+sys.path.insert(0, os.path.abspath("../"))
```
Comment (EmilHenriksen) [inline src/classification/extractor/low_level_data_extractor.py]: you said this wasn't needed anymore

File: src/__main__.py [hunk 7]
```diff
@@ -1,10 +1,12 @@
+#!/usr/local/bin/python3.6
+
 import sys
 import os
 from multiprocessing import Pool
 
 # Start of importing the utilities module
 # Needed due to the fact it imports a module that uses the utility module
-sys.path.insert(0, os.path.abspath("utilities/"))
+#sys.path.insert(0, os.path.abspath("utilities/"))
```
Comment (QuincySharp) [inline src/__main__.py]: Skal da bare slettes

File: src/classification/extractor/low_level_data_extractor.py [hunk 8]
```diff
@@ -1,17 +1,21 @@
...
-    extractor_path = get_absolute_path("utilities/ressources"
+    extractor_path = path_handler.get_absolute_path("utilities/ressources"
                                        + "/extractors/"
                                        + "streaming_extractor_music")
 
     command = '{} {} {}'.format(extractor_path, filename, output_file_path)
 
     subprocess.run(command, shell=True)
 
+    print("I am inside the extractor")
```
Comment (EmilHenriksen) [inline src/classification/extractor/low_level_data_extractor.py]: useless print

File: src/classification/extractor/low_level_data_extractor.py [hunk 9]
```diff
@@ -1,17 +1,21 @@
...
 
-from utilities.filehandler.handle_path import get_absolute_path
+sys.path.insert(0, os.path.abspath("../"))
+import utilities.filehandler.handle_path as path_handler 
 
 def make_low_level_data_file(filename, output_file_path):
-    extractor_path = get_absolute_path("utilities/ressources"
+    extractor_path = path_handler.get_absolute_path("utilities/ressources"
                                        + "/extractors/"
                                        + "streaming_extractor_music")
```
Comment (EmilHenriksen) [inline src/classification/extractor/low_level_data_extractor.py]: I would indent both lines one more so they line up

File: src/__main__.py [hunk 10]
```diff
@@ -72,3 +76,4 @@ def extract_and_save_data_from_songs_in_folder(folder_path):
     else:
         print("{} is not a folder or a path. Exiting...".format(arg))
         exit()
+"""
```
Comment (QuincySharp) [inline src/__main__.py]: Hvis det ikke bruges til noget skal det her også slettes

File: src/bpm/bpm_extractor.py [hunk 11]
```diff
@@ -1,8 +1,8 @@
 import sys
 import os
 from multiprocessing import Pool
```
Comment (QuincySharp) [inline src/bpm/bpm_extractor.py]: pep8 skal den her ned til de andre, tabulate og essentia iirc

File: src/rest_api/application.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+# Model is nested inside the song_fields model
+id_model = api.model('IdModel', {
+    'Release': fields.Integer(description='The release ID of the song', required=True),
+    'Side': fields.Integer(description='The side of the media', required=True),
+    'Track': fields.Integer(description='The tracknumber of the media', required=True),
+})
+
+song_fields = api.model('SongModel', {
+    'ID': fields.Nested(id_model, description='ID model of the song to analyze', required=False),
```
Comment (EmilHenriksen) [inline src/rest_api/application.py]: Too long pep 8

File: src/rest_api/application.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    'SourcePath': fields.String(description='The path of the song to analyze', required=True),
```
Comment (EmilHenriksen) [inline src/rest_api/application.py]: Too long

File: src/classification/__main__.py (hunk 3 above)
Comment (QuincySharp) [inline src/classification/__main__.py]: all these temp comments make no sense

File: src/rest_api/application.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+@api.route('/analyze_song')
+class AnalyzeSong(Resource):
+    @api.expect(song_fields)
+    def post(self):
+        data = request.get_json()
+        song_id = '{}-{}-{}'.format(
+            data["ID"]["Release"], data["ID"]["Side"],
+            data["ID"]["Track"]
+        )
+        # ../utilities/ressources/music/77245-1-1_Charles-Aznavour_Yesterday-when-i-was-young.wav
```
Comment (EmilHenriksen) [inline src/rest_api/application.py]: TOo long pep8

File: src/rest_api/t/test_application.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+def test_rest_server():
+    t = threading.Thread(target=start_rest)
+
+    t.start()
+
+    number_of_tries = 0
+    while number_of_tries < 10:
+        try:
+            r = requests.get("http://" + rest.hostURL + ":" + str(rest.hostPort) + rest.apiRoute)
```
Comment (EmilHenriksen) [inline src/rest_api/t/test_application.py]: Too long pep8

File: src/rest_api/t/test_application.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+            requests.get("http://" + rest.hostURL + ":" + str(rest.hostPort) + "/shutdown")
```
Comment (EmilHenriksen) [inline src/rest_api/t/test_application.py]: Too long pep8

File: src/classification/extractor/low_level_data_extractor.py (hunk 8 above)
Comment (QuincySharp) [inline src/classification/extractor/low_level_data_extractor.py]: Excuse me?

File: src/classification/extractor/low_level_data_extractor.py (hunk 6 above)
Comment (QuincySharp) [inline src/classification/extractor/low_level_data_extractor.py]: Is this still needed?

File: src/rest_api/application.py (hunk 14 above)
Comment (QuincySharp) [inline src/rest_api/application.py]: What is this comment explaining?

File: src/bpm/bpm_extractor.py (hunk 1 above)
Comment (Areedtz) [reply src/bpm/bpm_extractor.py]: Is the correct place. Tabulate is an external package installed by pip

File: src/classification/__main__.py (hunk 3 above)
Comment (Areedtz) [reply src/classification/__main__.py]: Removed the comment

File: docker-compose.yml (hunk 5 above)
Comment (Areedtz) [reply docker-compose.yml]: Under volumes. Indenting it further to make it more clear.

File: docker-compose.yml [hunk 17]
```diff
@@ -6,5 +6,7 @@ services:
...
-    command: tail -f /dev/null
+    ports:
+      - '127.0.0.1:80:1337'
```
[the remark this answers is not in the document]
Comment (Areedtz) [reply docker-compose.yml]: Changing indentation to be the same :)

File: src/classification/__main__.py (hunk 2 above)
Comment (Areedtz) [reply src/classification/__main__.py]: Moved up multiprocessing to fit pep8

File: src/bpm/bpm_extractor.py (hunk 11 above)
Comment (Areedtz) [reply src/bpm/bpm_extractor.py]: Multiprocessing is a default package in python, so it shouldn't be moved down.

File: src/classification/extractor/low_level_data_extractor.py (hunk 8 above)
Comment (Areedtz) [reply src/classification/extractor/low_level_data_extractor.py]: Sorry :)

File: src/__main__.py (hunk 10 above)
Comment (Areedtz) [reply src/__main__.py]: Removed :)

File: src/classification/extractor/low_level_data_extractor.py (hunk 6 above)
Comment (Areedtz) [reply src/classification/extractor/low_level_data_extractor.py]: I've changed this to be more correct.

File: src/classification/__main__.py (hunk 3 above)
Comment (Areedtz) [reply src/classification/__main__.py]: Removed the comments :)

File: src/rest_api/application.py (hunk 14 above)
Comment (Areedtz) [reply src/rest_api/application.py]: No longer necessary, so it has been removed.

File: src/rest_api/application.py (hunk 12 above)
Comment (Areedtz) [reply src/rest_api/application.py]: Shortened it.

File: src/rest_api/application.py (hunk 13 above)
Comment (Areedtz) [reply src/rest_api/application.py]: Shortened it.

File: src/rest_api/application.py (hunk 14 above)
Comment (Areedtz) [reply src/rest_api/application.py]: Removed it.

File: src/rest_api/t/test_application.py (hunk 15 above)
Comment (Areedtz) [reply src/rest_api/t/test_application.py]: Shortened it a bit.

File: src/rest_api/t/test_application.py (hunk 16 above)
Comment (Areedtz) [reply src/rest_api/t/test_application.py]: Shortened it a bit.

File: src/bpm/bpm_extractor.py (hunk 11 above)
Comment (EmilHenriksen) [reply src/bpm/bpm_extractor.py]: I mean tabulate should go up to multiprocessing

**Pull request merged**